pkg/reconciler/instance/certificate/reconciler.go

```diff
@@ -0,0 +1,341 @@
+package certificate
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+
+	"github.com/cloudnative-pg/machinery/pkg/fileutils"
+	"github.com/cloudnative-pg/machinery/pkg/log"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	apiv1 "github.com/cloudnative-pg/cloudnative-pg/api/v1"
+	"github.com/cloudnative-pg/cloudnative-pg/pkg/certs"
+	postgresSpec "github.com/cloudnative-pg/cloudnative-pg/pkg/postgres"
+)
+
+// Reconciler returns a certificate reconciler
+type Reconciler struct {
+	cli                      client.Client
+	serverCertificateHandler serverCertificateHandler
+}
+
+// NewReconciler creates a new certificate reconciler
+func NewReconciler(cli client.Client, serverHandler serverCertificateHandler) *Reconciler {
+	return &Reconciler{
+		cli:                      cli,
+		serverCertificateHandler: serverHandler,
+	}
+}
+
+type serverCertificateHandler interface {
+	SetServerCertificate(certificate *tls.Certificate)
+	GetServerCertificate() *tls.Certificate
```

Comment:
We should consider adding 'Lock' and 'Unlock' to enhance the server certificate refresh operation.